text/0000-mixed-shadow-mode.md

```diff
@@ -0,0 +1,175 @@
+---
+title: Mixed Shadow Mode
+status: DRAFTED
+created_at: 2021-05-07
+updated_at: YYYY-MM-DD
+pr: https://github.com/salesforce/lwc-rfcs/pull/49
+---
+
+# Mixed Shadow Mode
+
+## Introduction
+
+LWC implements a collection of polyfills that implement Shadow DOM features. These polyfills are
+collectively referred to as "synthetic Shadow DOM" and are published to npm under the
+`@lwc/synthetic-shadow` package. Synthetic shadow DOM provides a consistent experience for LWC web
+components across the Salesforce-supported browser matrix.
+
+Native Shadow DOM support has greatly increased since the start of the LWC project but applications
+continue to rely on the synthetic shadow polyfills to maintain backwards-compatibility. The removal
+of these polyfills can result in broken experiences due to dependencies on certain "escape hatches"
+that are not possible in a native Shadow DOM environment (e.g., instrumentation, styling, etc, which
+are discussed below).
+
+Applications that wish to use LWC components in a native shadow context simply need to omit the
+inclusion of the `@lwc/synthetic-shadow` polyfill; however, this is not a realistic option for
+large applications that cannot afford a rewrite.
+
+This proposal opens up an incremental migration path for applications moving to native Shadow DOM by
+allowing the usage of both native and synthetic Shadow DOM in the same document.
+
+## Detailed design
+
+Components will opt-in to this new feature and the default mode will be synthetic mode. From the
+component's perspective, there are two choices of shadow semantics: synthetic mode or native
+mode. The concept of a mixed mode should be transparent and the framework should guarantee this
+through integration testing, which is further discussed below.
+
+A component can signal that it prefers native Shadow DOM by setting a static property called
+`preferNativeShadow` to `true`.
+
+```js
+export default class extends LightningElement {
+    static preferNativeShadow = true;
+}
+```
+
+All polyfills applied by `@lwc/synthetic-shadow` will look to this property as a signal to invoke
+either the polyfilled API or the native browser API. If the browser does not support Shadow DOM
+(e.g., IE11), then LWC will fallback to synthetic mode regardless of which mode the component
+prefers.
+
+It is likely that components will need to know which mode it is currently operating in, due to
+observable differences between the two. Any indication of this by the framework should only be
+available to components that set the `preferNativeShadow` property to `true`.
+
+In terms of composition, a synthetic mode component can contain a native mode component, but the
+inverse is not allowed. Not only does this make things easier to reason about, but many existing

```

Comment:
What is the expected behavior when a parent component opt-in to native shadow while a child component is using synthetic shadow DOM? 

I originally thought that LWC should throw an error. This will cause issues when the flexipage component itself will be migrated to native shadow DOM since we can't ensure that all the components composing the flexipage are using native shadow as well. That said, migrating the flexipage component will certainly cause issues with Aura components as well.

Also I am wondering what would how this restriction would work when mixing light DOM components with shadow DOM components (native or not)

Reply:
> What is the expected behavior when a parent component opt-in to native shadow while a child component is using synthetic shadow DOM?

The current plan is to throw an exception during runtime.

> I originally thought that LWC should throw an error. This will cause issues when the flexipage component itself will be migrated to native shadow DOM since we can't ensure that all the components composing the flexipage are using native shadow as well. That said, migrating the flexipage component will certainly cause issues with Aura components as well.

If we do change our stance on this, it would be relaxing a restriction so I think we can deal with it once we have to.

> Also I am wondering what would how this restriction would work when mixing light DOM components with shadow DOM components (native or not)

Good question, I'll add it to the list of unresolved questions and think about it some more.

Reply:
FWIW yesterday we discussed light children inside of synthetic shadow parents and decided we wouldn't try to support style scoping for this. (I.e. `<style>`s will be inserted into the light child as-is; no effort is made to scope them to their shadow parent.) Not sure about other implications, but probably light DOM will work best with native shadow.